### PR TITLE
pipewire: 0.3.18 -> 0.3.21

### PR DESCRIPTION
**Pull request by jansol**
###### Motivation for this change

https://gitlab.freedesktop.org/pipewire/pipewire/-/releases/0.3.21
https://gitlab.freedesktop.org/pipewire/pipewire/-/releases/0.3.20
https://gitlab.freedesktop.org/pipewire/pipewire/-/releases/0.3.19

###### Things done

Initial attempt at generating config file in the new format using the upstream template as a base. Pulseaudio and jack (patchbays) seem to be working, everything else needs testing.

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/desktops/pipewire.nix [hunk 1]
```diff
@@ -46,12 +66,127 @@ in {
...
+            libpipewire-module-rtkit = {
+              args = {
+                nice.level = -11;
+                rt.prio = 20;
+                rt.time.soft = 200000;
+                rt.time.hard = 200000;
+              };
+              flags = "ifexists|nofail";
+            };
+            libpipewire-module-protocol-native = null;
```
Comment (eadwu) [inline nixos/modules/services/desktops/pipewire.nix:130]: This needs to be ordered in order for pipewire to work on my machine (thus the sortProperties in the generation), though I'm not too sure about how nice it looks with `libpipewire-module-protocol-native = { _priority = -100; _content = "null"; };` being what I have locally.

File: nixos/modules/services/desktops/pipewire.nix [hunk 2]
```diff
@@ -17,6 +17,26 @@ let
...
   '';
+
+  # Helpers for generating the pipewire JSON config file
+  mkSPAValueString = v:
+    if      builtins.isList       v then "[${lib.concatMapStringsSep " " mkSPAValueString v}]"
+    else if lib.types.attrs.check v then
+      "{${lib.concatStringsSep " " (mkSPAKeyValue { _SPA_DUP = v._type or "" == "_SPA_DUP"; } v)}}"
+    else lib.generators.mkValueStringDefault {} v;
+
+  mkSPAKeyValue = { _SPA_DUP ? false }: attrs: map (def: def.content) (lib.sortProperties
```
Comment (eadwu) [inline nixos/modules/services/desktops/pipewire.nix:28]: I couldn't get this working in any sort of elegant fashion, though the only part affected by this is the `objects` field for `pipewire.conf`.

**Review by eadwu — COMMENTED** [review]
I think it's still missing the `*-monitor` files, though overriding specific array values are a problem, though it can be mitigated by overrides if it is based on the order of the elements in the array.

Not quite sure about if the attribute sets are valid for the configuration, I only used a literal translation `"bluez5.msbc-support"`.

File: nixos/modules/services/desktops/pipewire.nix (hunk 2 above)
Comment (eadwu) [reply nixos/modules/services/desktops/pipewire.nix:28]: The simplified version that just gets rid of the `_SPA_DUP`

```nix
  mkSPAKeyValue = attrs: map (def: def.content) (
    lib.sortProperties
      (
        lib.mapAttrsToList
          (k: v: lib.mkOrder (v._priority or 1000) "${lib.escape [ "=" ] k} = ${mkSPAValueString (v._content or v)}")
          attrs
      )
  );
```

Comment (jansol) [reply nixos/modules/services/desktops/pipewire.nix:28]: Hmm how is that meant to be used? With the current `toSPAJSON = attrs: lib.concatStringsSep "\n" (mkSPAKeyValue {} attrs);` these just cause a "attempt to call something which is not a function but a list.

**Comment by jansol** [comment]
Now it builds and the config file looks ~~horrible~~ somewhat usable but the daemon fails to launch with: `pipewire[32692]: daemon 0x7ffde1ef2070: could not load module "libpipewire-module-metadata": Protocol error`. Need to figure out why that happens.

There are also a couple of config files for the media session that need to be generated...

**Comment by eadwu** [comment]
I'm pretty sure that's because of protocol-native needs to be loaded first. My local config in a gist, https://gist.github.com/eadwu/321fb3b45e0790424ba88488f7407b85.

**Comment by jansol** [comment]
Ah, that's what you meant by that. Let's see how adapting that gist of yours works out.

**Comment by ashkitten** [comment]
any progress on this?

**Comment by jansol** [comment]
Been busy with work and other stuff, planning to get back to this PR in the evening.

**Comment by collares** [comment]
This is more of a suggestion for future PRs, but I found #111076 interesting. For Mattermost (completely unrelated to pipewire), the corresponding NixOS module parses the default JSON config file provided by the package, instead of duplicating most of the data in it. A similar idea could be used in pipewire. A possible downside of doing this is explained in the linked issue.

**Comment by jtojnar** [comment]
Using IFD in the module system will not sit right with Hydra, which makes that method a no-go for module that is enabled by default in desktop tests.

**Comment by collares** [comment]
@jtojnar Would it still be a problem if the default config files were copied to `nixos/modules/services/desktops` instead of fetched from the Nix store? It would still be somewhat easier to sync with changes in the default config files.

**Comment by jtojnar** [comment]
That should work. We do something similar in fwupd (copy the data to `passthu` and then check consistency in `passthru.tests`): https://github.com/NixOS/nixpkgs/blob/aa48e205a2e0516ff4b96c79eeb666aaaccd418c/pkgs/os-specific/linux/firmware/fwupd/default.nix#L275-L330

**Comment by jansol** [comment]
One problem though: the default config is generated with absolute paths to the pipewire and pipewire-media-session binaries at build time...

**Comment by jansol** [comment]
Yay, it's making noises again. Even with over bluetooth (which is a first for me with pipewire)! Someone who actually knows what they are doing should really look it over though, there is now way this is passing tests as is. Also I didn't really get how the passthru stuff is supposed to work.

File: nixos/modules/services/desktops/pipewire.nix [hunk 3]
```diff
@@ -46,32 +98,112 @@ in {
...
-        example = literalExample ''"''${pipewire}/bin/pipewire-media-session"'';
+        type = types.str;
+        default = "${cfg.package}/bin/pipewire-media-session";
+        example = literalExample ''"\$\{pipewire\}/bin/pipewire-media-session"'';
         description = ''
           Path to the pipewire session manager executable.
         '';
       };
 
-      sessionManagerArguments = mkOption {
```
[the remark this answers is not in the document]
Comment (jansol) [reply nixos/modules/services/desktops/pipewire.nix:66]: I was thinking of adding an extraEtcFiles instead since that's the way to configure pipewire-media-session now. But for other session managers it makes sense to keep this around I guess.

[the remark this answers is not in the document]
Comment (ashkitten) [reply nixos/modules/services/desktops/pipewire.nix:66]: maybe `settings` as per https://github.com/NixOS/rfcs/pull/42?

[the remark this answers is not in the document]
Comment (jansol) [reply nixos/modules/services/desktops/pipewire.nix:66]: The problem is that depending on the session manager there can be an arbitrary number of config files and their names aren't known ahead of time. The current solution allows specifying arbitrary files and if the `toSPAJSON` function is made public (just looking for the right place for it) it would allow specifying settings for `pipewire-media-session` mostly in this way, just like is done with pipewire.conf (except for having to manually call the fn). But a fancy config generation setup can be done separately too.

**Comment by MetaDark** [comment]
Thanks for this @jansol. I tried out this build because I got a new pair of headphones that uses AAC (now available in 0.3.19), and the audio works great! Although, I'm having problems configuring things through pavucontrol:
```
Feb 02 14:03:57 atlantis pipewire-pulse[2105]: pulse-server 0x9f0bd0: [PulseAudio Volume Control] ERROR command:96 (SET_SINK_PORT) tag:35 error:1 (Permission denied)
Feb 02 14:03:57 atlantis pipewire-pulse[2105]: pulse-server 0x9f0bd0: [PulseAudio Volume Control] ERROR command:97 (SET_SOURCE_PORT) tag:36 error:1 (Permission denied)
Feb 02 14:03:57 atlantis pipewire-pulse[2105]: pulse-server 0x9f0bd0: [PulseAudio Volume Control] ERROR command:96 (SET_SINK_PORT) tag:37 error:1 (Permission denied)
Feb 02 14:04:03 atlantis pipewire-pulse[2105]: pulse-server 0x9f0bd0: [PulseAudio Volume Control] ERROR command:36 (SET_SINK_VOLUME) tag:38 error:1 (Permission denied)
```

From what I understand, those commands should only be modifying files under `XDG_RUNTIME_DIR`, so I'm not sure why it's failing now.

**Comment by collares** [comment]
Speaking of AAC, should `fdk_aac` be added to `buildInputs`?

**Comment by jansol** [comment]
> Speaking of AAC, should fdk_aac be added to buildInputs?

Done.

As for the config stuff, the current solution is ugly and terrible in many ways but should allow for easy(ish) prototyping something better. However figuring out the fine details of nix isn't really in my interests and I'd like to get back to following upstream releases ASAP so unless there are major complaints about the current state of this PR I'd like to merge it as is and let someone more knowledgeable figure out a nice way to configure pipewire and its session managers.

**Comment by jansol** [comment]
> I'm having problems configuring things through pavucontrol:

Can confirm the same problem here. Might be an upstream issue though, considering I saw someone complain about a similar problem on fedora.

**Comment by eadwu** [comment]
Codec switching is on master. Can be controlled through pavucontrol as like `pulseaudio-modules-bt`.

**Comment by ashkitten** [comment]
testing this the other day, i was getting more audio dropouts (cascading xruns) than usual when scrubbing a video in mpv.. i'm told that i should increase alsa.headroom, but it's not clear how to do that with the nixos module here

**Comment by jansol** [comment]
TL;DR: changing any configuration atm basically means replacing the relevant config file either by writing its contents to `services.pipewire.config` or by adding an entry to `services.pipewire.sessionManagerEtcFiles` i.e.:
```
services.pipewire.sessionManagerEtcFiles = {
  "pipewire/media-session.d/alsa-monitor.conf" = { text = "#generate file contents somehow here"; };
};
```

It's super ugly but I'm not familiar enough with the nixos module system to come up with anything better right now and don't really have time nor motivation to figure out something fancy. The whole affair is made more complicated by the fact that I don't know how the configuration for other session managers would look. Likely the most elegant solution would be to delegate session manager configuration to session manager-specific modules. This would mean splitting pipewire-session-manager into its own package and adding a separate module for it.

**Comment by ashkitten** [comment]
does session manager config need to be generic? i assume that other session managers will have their own modules and be configured separately; either way, since pipewire-media-session is the only session manager currently, i think it's okay to treat it specially.

**Comment by jansol** [comment]
I'd sure hope that session managers are configurable *somehow*. But it's also past 5am and I'm really not going to think about it now.

**Comment by ashkitten** [comment]
a session manager is a separate program that can be executed at pipewire startup. pipewire-media-session is part of pipewire, so it is configured like this. wireplumber for instance has configuration files that live in /etc/wireplumber, so no need to worry about that at all.

**Comment by collares** [comment]
0.3.21 is now out. Hopefully it's just a matter of replacing the config files.

**Comment by jansol** [comment]
Okay this looks a lot nicer. Whoever wants to try, here is what I'm using currently:
```
  security.rtkit.enable = true;
  services.pipewire = {
    enable = true;
    alsa.enable = true;
    alsa.support32Bit = true;
    pulse.enable = true;
    jack.enable = true;

    # use the example session manager
    pwms.enable = true;
  };
```

For customizing the configuration there are attrsets that get converted to pipewire and pipewire-media-session config files:
`services.pipewire.config`
`services.pipewire.pwms.config`
`services.pipewire.pwms.alsaMonitorConfig`
`services.pipewire.pwms.bluezMonitorConfig`
`services.pipewire.pwms.v4l2MonitorConfig`

**Comment by Mic92** [comment]
>     # use the example session manager
>     pwms.enable = true;

I just tested the new upgrade and it looks like this session manager is now mandatory or it does not work (i.e. no devices are shown in pulseaudio). Therefore I would enable it by default.

File: nixos/modules/services/desktops/pipewire/pipewire-media-session.nix [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+
+      bluezMonitorConfig = mkOption {
+        type = types.attrs;
+        description = ''
+          Configuration for the bluez5 monitor.
+        '';
+        default = {
+          # bluez-monitor config file
+          properties = {
+            # msbc is not expected to work on all headset + adapter combinations.
```
Comment (Mic92) [inline nixos/modules/services/desktops/pipewire/pipewire-media-session.nix:199]: What do we do with all the commented out configuration here? This looks like it could easily get out of date.

**Comment by Mic92** [comment]
Have you tested headphones? I just tried connecting them, but pipewire does no longer switch to them.

**Comment by jansol** [comment]
Bluetooth ones work fine, both for input and output. Wired works with the jacks on the motherboard IO panel but not on the chassis front panel. That said, I'm not even sure if I wired that one up in the first place since I always keep my headphones attached to the USB soundcard. Automatic route switching is a bit flaky at times but switching the default output in pavucontrol puts things where they belong.

File: nixos/modules/services/desktops/pipewire/pipewire-media-session.nix (hunk 4 above)
Comment (jansol) [reply nixos/modules/services/desktops/pipewire/pipewire-media-session.nix:199]: I think that was what [this comment](https://github.com/NixOS/nixpkgs/pull/110615#issuecomment-769980318) and its immediate replies were getting at.

> looks like this session manager is now mandatory or it does not work

More specifically 'a' session manager is mandatory (always has been) but this is the only one that is packaged for now. Does `mkEnableOption` support setting the option to true by default?

**Comment by ashkitten** [comment]
automatic port/profile switching seems to now be disabled by default at least in your pipewire-media-session config here

File: nixos/modules/services/desktops/pipewire/pipewire.nix [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        '';
+      };
+
+      config = mkOption {
+        type = types.attrs;
+        description = ''
+          Configuration for the pipewire daemon.
+        '';
+        default = {
+          properties = {
```
Comment (collares) [inline nixos/modules/services/desktops/pipewire/pipewire.nix:73]: This is somewhat different from `pipewire.conf.in`, which only sets `link.max-buffers = 16` and does not touch the quantum.

File: nixos/modules/services/desktops/pipewire/pipewire.nix [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+                restricted = [];
+                force = "flatpak";
+              };
+            };
+            libpipewire-module-adapter = "null";
+            libpipewire-module-link-factory = "null";
+            libpipewire-module-session-manager = "null";
+          };
+
+          objects = {
```
Comment (collares) [inline nixos/modules/services/desktops/pipewire/pipewire.nix:154]: spa-node-factory is missing here, perhaps? The comments in `pipewire.conf.in` say this affects JACK clients.

Comment (eadwu) [reply nixos/modules/services/desktops/pipewire/pipewire.nix:154]: The format allows for duplicate keys, which is hard to represent through Nix's structured attributes (without messing implementing it in a non-nix friendly way).

File: nixos/modules/services/desktops/pipewire/pipewire-media-session.nix (hunk 4 above)
Comment (Mic92) [reply nixos/modules/services/desktops/pipewire/pipewire-media-session.nix:199]: @jansol No, not directly but you can use lib.mkOption with type lib.bool.

**Comment by Mic92** [comment]
> automatic port/profile switching seems to now be disabled by default at least in your pipewire-media-session config here

I remember it work. Could we switch it back on?

**Comment by ashkitten** [comment]
> > automatic port/profile switching seems to now be disabled by default at least in your pipewire-media-session config here
> 
> I remember it work. Could we switch it back on?

this is the new default. you can always change it for yourself, but my personal opinion is that we should stick to the defaults at least for the moment for simplicity.

**Comment by Mic92** [comment]
> > > automatic port/profile switching seems to now be disabled by default at least in your pipewire-media-session config here
> > 
> > 
> > I remember it work. Could we switch it back on?
> 
> this is the new default. you can always change it for yourself, but my personal opinion is that we should stick to the defaults at least for the moment for simplicity.

Ok. Sound good to me.

File: nixos/modules/services/desktops/pipewire/pipewire-media-session.nix [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+            # (Sony WH-1000XM3) are not working with both hsp_ag and hfp_ag
+            # enabled, disable either hsp_ag or hfp_ag to work around it.
+            #
+            # Supported headset roles: hsp_hs (HSP Headset),
+            #                          hsp_ag (HSP Audio Gateway),
+            #                          hfp_ag (HFP Audio Gateway)
+            #bluez5.headset-roles = [ hsp_hs hsp_ag hfp_ag ]
+
+            # Enabled A2DP codecs (default: all)
+            #bluez5.codecs = [ sbc aac ldac aptx aptx_hd ]
```
Comment (Mic92) [inline nixos/modules/services/desktops/pipewire/pipewire-media-session.nix:214]: Should the commented out code not be at least valid nix code?

File: nixos/modules/services/desktops/pipewire/pipewire-media-session.nix [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+              # actions can update properties on the matched object.
+              update-props = {
+                api.alsa.use-acp = true;
+                #api.alsa.use-ucm = true
+                #api.alsa.soft-mixer = false
+                #api.alsa.ignore-dB = false
+                #device.profile-set = "profileset-name"
+                #device.profile = "default profile name"
+                api.acp.auto-profile = false;
+                api.acp.auto-port = false;
```
Comment (Mic92) [inline nixos/modules/services/desktops/pipewire/pipewire-media-session.nix:150]: How would I override these values from a nixos module? Also is `auto-port` and `auto-profile` the options I need to set to make pipewire switch to headphone once it is plugged in?

**Comment by kamushadenes** [comment]
Hi, what remains to be done for this to be merged?

I'm hoping to solve a bizarre mic crackling issue by upgrading to `>= 0.3.20` as per https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/475.

File: nixos/modules/services/desktops/pipewire/pipewire-media-session.nix (hunk 8 above)
Comment (jansol) [reply nixos/modules/services/desktops/pipewire/pipewire-media-session.nix:150]: See updated OP in #102547 .

I'm not familiar with any of the options besides the two shown in the example.

File: nixos/modules/services/desktops/pipewire/pipewire-media-session.nix (hunk 7 above)
Comment (jansol) [reply nixos/modules/services/desktops/pipewire/pipewire-media-session.nix:214]: Certainly wouldn't hurt. On the other hand the whole thing should be redone and probably autogenerated from upstream config files. Whenever someone has the time and knowledge to implement that.

**Comment by collares** [comment]
One more nitpick (sorry!): Can we use `media-session` instead of `pwms`? Might be a bit easier to discover since `services.pipewire.media-session` sort of matches `/etc/pipewire/media-session.d/`

**Comment by Mic92** [comment]
Don't yet why but `nix-build nixos/tests/installed-tests/default.nix -A pipewire` fails with infinite recursion. This is also what ofborg complains about. Otherwise the PR now looks good to me.

File: nixos/modules/services/desktops/pipewire/pipewire.nix (hunk 5 above)
Comment (Mic92) [reply nixos/modules/services/desktops/pipewire/pipewire.nix:73]: cc @jansol is there a motivation behind this?

File: nixos/modules/services/desktops/pipewire.nix (hunk 1 above)
Comment (Mic92) [reply nixos/modules/services/desktops/pipewire.nix]: Is this addressed now? Do we care?

File: nixos/modules/services/desktops/pipewire.nix (hunk 2 above)
Comment (Mic92) [reply nixos/modules/services/desktops/pipewire.nix]: How about this?

**Comment by eadwu** [comment]
Next release will have another wave of configuration files if master stays the same. Format will be the same, but there will be a lot more files.

**Comment by Mic92** [comment]
> Next release will have another wave of configuration files if master stays the same. Format will be the same, but there will be a lot more files.

Mhm. Is there a better way to provide those without having to provide all their content in our module? This becomes a maintenance nightmare otherwise.

**Comment by Mic92** [comment]
cc @wtay @pv, we want to make pipewire configurable in NixOS through NixOS options. Basically we convert our own nix data structures to pipewire configuration format (see https://github.com/NixOS/nixpkgs/pull/110615/files#diff-6c167307e84d42c088d640f1859ad87e6c8cb15c1ce2353361eac78b5351027eR58). Do you see a way within the format that would allow us to use upstream defaults while still allowing all configuration items to be imported without importing all upstream defaults in our module?

**Comment by collares** [comment]
As I understand it, a valid JSON file should work as a configuration file, but the SPA JSON parser accepts more than just well-formed JSON on purpose (to make the configuration file less cluttered?). So outputting a config file should be doable with standard tools, but we would need to port https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/spa/include/spa/utils/json.h to Nix (or have an update script that preprocesses stuff) be able to parse the sample config files.

**Comment by Mic92** [comment]
> As I understand it, a valid JSON file should work as a configuration file, but the SPA JSON parser accepts more than just well-formed JSON on purpose (to make the configuration file less cluttered?). So outputting a config file should be doable with standard tools, but we would need to port [gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/spa/include/spa/utils/json.h](https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/spa/include/spa/utils/json.h) to Nix (or have an update script that preprocesses stuff) be able to parse the sample config files.

Yes. Ideally something that does it at runtime.

**Comment by Mic92** [comment]
> As I understand it, a valid JSON file should work as a configuration file, but the SPA JSON parser accepts more than just well-formed JSON on purpose (to make the configuration file less cluttered?). So outputting a config file should be doable with standard tools, but we would need to port [gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/spa/include/spa/utils/json.h](https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/spa/include/spa/utils/json.h) to Nix (or have an update script that preprocesses stuff) be able to parse the sample config files.

There you go: https://github.com/nix-community/pipewire-to-json

File: nixos/modules/services/desktops/pipewire.nix (hunk 2 above)
Comment (eadwu) [reply nixos/modules/services/desktops/pipewire.nix]: If we're making converting the JSON to Nix sets before reconverting that to SPA JSON then more or less you'll still need parts of this. `sortProperties` will definitely be needed because of the `load-module` ordering dependency.

**Comment by wtay** [comment]
> > As I understand it, a valid JSON file should work as a configuration file, but the SPA JSON parser accepts more than just well-formed JSON on purpose (to make the configuration file less cluttered?). So outputting a config file should be doable with standard tools, but we would need to port [gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/spa/include/spa/utils/json.h](https://gitlab.freedesktop.org/pipewire/pipewire/-/blob/master/spa/include/spa/utils/json.h) to Nix (or have an update script that preprocesses stuff) be able to parse the sample config files.
> 
> There you go: https://github.com/nix-community/pipewire-to-json

There is now also a canonical converter in PipeWire: https://gitlab.freedesktop.org/pipewire/pipewire/-/commit/3c9996aa781c3cf3623547dbddad4772196ae391

File: nixos/modules/services/desktops/pipewire.nix (hunk 2 above)
Comment (eadwu) [reply nixos/modules/services/desktops/pipewire.nix]: Also get's messy if there are multiple `objects` with the same factory since duplicate keys aren't allowed in Nix sets.

**Comment by collares** [comment]
Would it be worth getting this merged as-is and waiting until 0.3.22 to clean up the configuration file handling (so the new `spa-json-dump` utility created by wtay can be used)?

**Comment by Mic92** [comment]
> Would it be worth getting this merged as-is and waiting until 0.3.22 to clean up the configuration file handling (so the new `spa-json-dump` utility created by wtay can be used)?

I am fine either way, before we can merge this the evaluation error needs to be fixed so. Do you think it would change the module interface a lot?

File: pkgs/development/libraries/pipewire/default.nix [hunk 9]
```diff
@@ -36,111 +39,117 @@ let
...
+    doCheck = true;
+
+    postInstall = ''
+      moveToOutput "share/systemd/user/pipewire-pulse.*" "$pulse"
+      moveToOutput "lib/systemd/user/pipewire-pulse.*" "$pulse"
+      moveToOutput "bin/pipewire-pulse" "$pulse"
+      moveToOutput "bin/pipewire-media-session" "$mediaSession"
+    '';
+
+    passthru = {
```
Comment (collares) [inline pkgs/development/libraries/pipewire/default.nix:131]:
```suggestion
    passthru.tests = {
```
fixes the `installed-tests` evaluation error for me.

Comment (jansol) [reply pkgs/development/libraries/pipewire/default.nix]: Thanks for spotting this! Must have slipped through while I tested stuff that involved changing this section.

**Comment by cawilliamson** [comment]
Personally I'd like to see this merged as soon as possible - for me it's meaning I can't fully switch to pipewire because it breaks some apps which ignore the default output defined.

It looks like the fix for this is in this newer version. :)

**Comment by Izorkin** [comment]
@jansol @Mic92 after this update on my home server automatic activated this service.

**Comment by Kazimazi** [comment]
my pipewire refuses to start, `/etc/pipewire/pipewire.conf` is missing
some `journalctl` info:
```
Feb 18 23:13:01 amdboi pipewire[1590]: failed to parse config: /nix/store/43884baaflihk674g4zddg1zhyxlam6a-pipewire.conf:2: Command "modules" does not exist
Feb 18 23:13:01 amdboi systemd[1582]: pipewire.service: Main process exited, code=exited, status=255/EXCEPTION
Feb 18 23:13:01 amdboi systemd[1582]: pipewire.service: Failed with result 'exit-code'.
Feb 18 23:13:01 amdboi systemd[1582]: pipewire.service: Scheduled restart job, restart counter is at 1.
Feb 18 23:13:01 amdboi systemd[1582]: Stopped Multimedia Service.
Feb 18 23:13:01 amdboi systemd[1582]: Started Multimedia Service.
Feb 18 23:13:01 amdboi pipewire[1807]: failed to parse config: /nix/store/43884baaflihk674g4zddg1zhyxlam6a-pipewire.conf:2: Command "modules" does not exist
...
Feb 18 23:13:02 amdboi systemd[1582]: pipewire.service: Scheduled restart job, restart counter is at 5.
Feb 18 23:13:02 amdboi systemd[1582]: Stopped Multimedia Service.
Feb 18 23:13:02 amdboi systemd[1582]: pipewire.service: Start request repeated too quickly.
Feb 18 23:13:02 amdboi systemd[1582]: pipewire.service: Failed with result 'exit-code'.
Feb 18 23:13:02 amdboi systemd[1582]: Failed to start Multimedia Service.
Feb 18 23:13:02 amdboi systemd[1582]: pipewire.socket: Failed with result 'service-start-limit-hit'.
Feb 18 23:13:06 amdboi pipewire[1888]: daemon 0x7ffcffca98f0: error loading config '/etc/pipewire/pipewire.conf': No such file or directory
Feb 18 23:13:06 amdboi pipewire[1888]: failed to load config: No such file or directory
```

**Comment by collares** [comment]
`/etc/pipewire/pipewire.conf` is not supposed to exist, since the `PIPEWIRE_CONFIG_FILE` env variable (defined for the systemd user service) points somewhere else. Do you see something weird in `/nix/store/43884baaflihk674g4zddg1zhyxlam6a-pipewire.conf`?

**Comment by Kazimazi** [comment]
I'm not sure what it supposed to look like, that's what I've got:
<details>
<summary>pipewire.conf</summary>
exec = {/nix/store/kq5ywwjjg7xyjplprirx5m5ijf1vav58-pipewire-0.3.21-mediaSession/bin/pipewire-media-session = {args = ""}}
modules = {libpipewire-module-protocol-native = null libpipewire-module-access = {args = {access = {allowed = [/nix/store/kq5ywwjjg7xyjplprirx5m5ijf1vav58-pipewire-0.3.21-mediaSession/bin/pipewire-media-session] force = flatpak rejected = [] restricted = []}}} libpipewire-module-adapter = null libpipewire-module-client-device = null libpipewire-module-client-node = null libpipewire-module-link-factory = null libpipewire-module-metadata = null libpipewire-module-portal = null libpipewire-module-profiler = null libpipewire-module-rtkit = {args = {} flags = ifexists|nofail} libpipewire-module-session-manager = null libpipewire-module-spa-device-factory = null libpipewire-module-spa-node-factory = null}
objects = {}
properties = {link.max-buffers = 16}
spa-libs = {api.alsa.* = alsa/libspa-alsa api.bluez5.* = bluez5/libspa-bluez5 api.jack.* = jack/libspa-jack api.libcamera.* = libcamera/libspa-libcamera api.v4l2.* = v4l2/libspa-v4l2 api.vulkan.* = vulkan/libspa-vulkan audio.convert* = audioconvert/libspa-audioconvert support.* = support/libspa-support}
</details>

and that's how I configured it:
``` nix
{ config, pkgs, ... }: {
  security.rtkit.enable = true;
  nixpkgs.config.pulseaudio = true;
  hardware.pulseaudio.enable = pkgs.lib.mkForce false;
  services.pipewire = {
    enable = true;
    package = pkgs.master.pipewire;
    alsa.enable = true;
    pulse.enable = true;
    jack.enable = true;
  };
}

```

**Comment by Technical27** [comment]
@Kazimazi pipewire's config got a massive change, you are using the module from stable (which expects 0.3.18), but pipewire 0.3.20 can't read the old config format.

**Comment by collares** [comment]
Just to be sure: Does your config file really not contain a line break before "module"?

**Comment by Kazimazi** [comment]
It might be something trivial, but I can't wrap my head around it. How do I use the new module?

**Comment by collares** [comment]
What is "pkgs.master" on your config? I don't have this attribute on my local config. Could it be pointing to an old package set? I think you're using the new module but "pkgs.master.pipewire" is an old package.

**Comment by Kazimazi** [comment]
It's an overlay of master branch of nixpkgs, which is updated.

File: nixos/modules/services/desktops/pipewire/pipewire-media-session.nix [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+  meta = {
+    maintainers = teams.freedesktop.members;
+  };
+
+  ###### interface
+  options = {
+    services.pipewire.media-session = {
+      enable = mkOption {
+        type = types.bool;
+        default = true;
```
Comment (betaboon) [inline nixos/modules/services/desktops/pipewire/pipewire-media-session.nix:40]: @jansol this seems like a bad idea. i just ran into a problem when building a system and pipewire being pulled in due to this

**Comment by Kazimazi** [comment]
I think I've got my issue resolved... there was a symlink to a non-existent /nix/store/...pipewire3.18 in ~/.config/systemd/user and it was conflicting with the new one. Don't know when or why it got there so I deleted it.

File: nixos/modules/services/desktops/pipewire/pipewire-media-session.nix (hunk 10 above)
Comment (jansol) [reply nixos/modules/services/desktops/pipewire/pipewire-media-session.nix:40]: I'm open to better suggestions / instructions. The idea was that since pipewire-media-session is currently the only packaged session manager it has to be enabled by default. Of course it is not relevant when pipewire itself is not enabled.

Comment (minijackson) [reply nixos/modules/services/desktops/pipewire/pipewire-media-session.nix:40]: Would it be possible to set the default to `config.services.pipewire.enabled`? Since it wouldn't be relevant otherwise as you said.

Comment (betaboon) [reply nixos/modules/services/desktops/pipewire/pipewire-media-session.nix:40]: @minijackson suggestion sounds reasonable to me. that would prevent pulling in pipewire into a system when not even being used.

Comment (Mic92) [reply nixos/modules/services/desktops/pipewire/pipewire-media-session.nix:40]: fixed in https://github.com/NixOS/nixpkgs/pull/113772

**Comment by d4g** [comment]
> I think I've got my issue resolved... there was a symlink to a non-existent /nix/store/...pipewire3.18 in ~/.config/systemd/user and it was conflicting with the new one. Don't know when or why it got there so I deleted it.

I ran into the same issue with #115492 .
This is pretty weird and an issue that more users may run into when updating. Probably something to consider, that somehow those files seem to have been created on an earlier version.

```
.config/systemd/user:
./pipewire.socket
./default.target.wants
./default.target.wants/pipewire.service
./pipewire.service
./sockets.target.wants
./sockets.target.wants/pipewire.socket
```

```
pipewire.service -> /nix/store/sl82a2fpsyn7ihjp0lp67p7f4k5sp5dh-pipewire-0.3.18/lib/systemd/user/pipewire.service
pipewire.socket -> /nix/store/sl82a2fpsyn7ihjp0lp67p7f4k5sp5dh-pipewire-0.3.18/lib/systemd/user/pipewire.socket
```
